src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs

```diff
@@ -48,45 +54,62 @@ public partial class HubConnection
         /// </summary>
         public TimeSpan ServerTimeout { get; set; } = DefaultServerTimeout;
 
-        public HubConnection(IConnection connection, IHubProtocol protocol, ILoggerFactory loggerFactory)
+        public HubConnection(Func<IConnection> connectionFactory, IHubProtocol protocol) : this(connectionFactory, protocol, NullLoggerFactory.Instance)
+        {
+        }
+
+        public HubConnection(Func<IConnection> connectionFactory, IHubProtocol protocol, ILoggerFactory loggerFactory)
         {
-            if (connection == null)
+            if (connectionFactory == null)
             {
-                throw new ArgumentNullException(nameof(connection));
+                throw new ArgumentNullException(nameof(connectionFactory));
             }
 
             if (protocol == null)
             {
                 throw new ArgumentNullException(nameof(protocol));
             }
 
-            _connection = connection;
+            _connectionFactory = connectionFactory;
             _binder = new HubBinder(this);
             _protocol = protocol;
             _loggerFactory = loggerFactory ?? NullLoggerFactory.Instance;
             _logger = _loggerFactory.CreateLogger<HubConnection>();
-            _connection.OnReceived((data, state) => ((HubConnection)state).OnDataReceivedAsync(data), this);
-            _connection.Closed += Shutdown;
 
             // Create the timer for timeout, but disabled by default (we enable it when started).
-            _timeoutTimer = new Timer(state => ((HubConnection)state).TimeoutElapsed(), this, Timeout.Infinite, Timeout.Infinite);
+            _timeoutTimer = new Timer(state => _ = ((HubConnection)state).TimeoutElapsed(), this, Timeout.Infinite, Timeout.Infinite);
         }
 
         public async Task StartAsync()
         {
+            CheckDisposed();
+
+            await _connectionLock.WaitAsync();
```

Comment:
This await will resume back on the sync context. ConfigureAwait(false)? Or move connection lock inside StartAsyncCore.

Reply:
Yeah, I'll move the locks inside the ForceAsync logic.